### DISCUSSION TITLE
Is there a way to import non-trivial functions from one Blocks program into another one?

### DISCUSSION THREAD

**Discussion by ggramlich** [pybricks/support#1344] [General]
I was hoping that we could define some functions with a drivebase, in one Blocks program (library.py) and then import it into another one (mission.py).
Syntactically this is possible, but as both programs need to have a setup for the drivebase, the code in library.py is executed on import and mission.py cannot define the motors on its own (error message that the port is already in use). It is possible to import motors and drivebase, from library.py, but then mission.py does not know their type and you cannot use them.

I do understand how to build library.py in Python. Do you see any way to support building a library in Blocks?
I thought, perhaps just something like this pattern is missing, but did not think this through:

```
if __name__ == "__main__":
    main()

```

**Comment by laurensvalk** [comment]
You can make any function with blocks or Python and mix them as you like.

Just as with Python, it's best to do the setup in only one file.

You can still make generic motor/drivebase functions by passing in the object. You can find the objects on the variables tab.

**Comment by ggramlich** [comment]
Thank you so much @laurensvalk. I once saw the devices under the variables tab, but then completely forgot them.
For anybody else who ist struggling with that, here is an example:

The robot starts orthogonal to a black line, then drives straight up to it. He crosses the line to have the axle on it and turns to the right to then start line following.

![pybricks_issue_import_library](https://github.com/pybricks/support/assets/69276/18d30b1b-6326-46da-aeba-6e429c261913)

![pybricks_issue_import_mission](https://github.com/pybricks/support/assets/69276/5051b4a0-5f02-4f65-a059-e4ec78cac314)

**Comment by laurensvalk** [comment]
Hi @ggramlich . Nice! I had the opposite in mind but yes, yours works too. I'll add my example for reference.

-------------------------

I was imagining a library that is usable independently. This could have reusable functions without any setup. Instead, the setup would happen in the application.

For example, here is a library with a countdown display and polygon drive function:

![image](https://github.com/pybricks/support/assets/12326241/5eba8818-f729-412a-8918-82bbee5224c1)

And the application would pass in its devices to use the library functions:

![image](https://github.com/pybricks/support/assets/12326241/bd167a79-951a-4855-bc9a-a8fcc1fc1c2c)

**Comment by ggramlich** [comment]
@laurensvalk That makes perfect sense to me. Thanks for adding the example. We will see, how we are going to use it.
